cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java

```diff
@@ -0,0 +1,264 @@
+
+
+      for (String inputStage : sortedInputs) {
+        List<Operation> operations = stageOperations.get(inputStage);
+        for (Operation operation : operations) {
+          String origin = prefixedOperationName(inputStage, operation.getName());
+          List<String> outputsOfOperation = new ArrayList<>();
+          switch (operation.getType()) {
+            case READ:
+              ReadOperation read = (ReadOperation) operation;
+              outputsOfOperation.addAll(read.getOutputFields());
+              break;
+            case TRANSFORM:
+              TransformOperation transform = (TransformOperation) operation;
+              outputsOfOperation.addAll(transform.getOutputFields());
+              break;
+            default:
+          }
+          for (String output : outputsOfOperation) {
+            inputFields.add(InputField.of(origin, output));
+          }
+        }
+      }
+
+      List<String> outputs = new ArrayList<>();
+      Set<String> alreadyAddedOutputs = new HashSet<>();
+      for (InputField inputField : inputFields) {
+        if (alreadyAddedOutputs.contains(inputField.getName())) {
+          continue;
+        }
+        outputs.add(inputField.getName());
+        alreadyAddedOutputs.add(inputField.getName());
+      }
+      co.cask.cdap.api.lineage.field.TransformOperation merge
+              = new co.cask.cdap.api.lineage.field.TransformOperation(mergeOperationName, mergeDescription,
+                                                                      inputFields, outputs);
+      processedOperations.add(merge);
+    }
+  }
+
+  /**
+   * Create {@link InputField}s from field names which acts as an input to the operation. Creating
+   * InputField requires origin; the name of the operation which created the field. To figure out the
+   * origin, we traverse in the reverse direction from the current stage until we reach source. While
+   * traversing, for each stage we check if the field occurs in any of the output operations recorded
+   * by that stage, if so we have found the origin.
+   *
+   * @param fields the List of input field names for an operation for which InputFields to be created
+   * @param currentStage name of the stage which recorded the operation
+   * @param parents topological ordered list of stages till the current stage
+   * @return List of InputFields
+   */
+  private List<InputField> createInputFields(List<String> fields, String currentStage, List<String> parents) {
+    List<InputField> result = new ArrayList<>();
+    for (String field : fields) {
+      ListIterator<String> branchIterator = parents.listIterator(parents.size());
+      boolean foundOrigin = false;
+      while (branchIterator.hasPrevious() && !foundOrigin) {
+        String stage = branchIterator.previous();
+        // check if the current field is output by any one of operation created by current stage
+        List<Operation> operations = stageOperations.get(stage);
```

Comment:
it would be good to add a comment why it is important to iterate over the operations in reverse order. It took me a while to realize why.

Reply:
simplified this code by maintaining additional map of stage-><field_name, origin>